module/builder/collection/builder.go

```diff
@@ -60,277 +53,277 @@ func NewBuilder(
 		apply(&b.config)
 	}
 
-	return &b
+	// sanity check config
+	if b.config.ExpiryBuffer >= flow.DefaultTransactionExpiry {
+		return nil, fmt.Errorf("invalid configured expiry buffer exceeds tx expiry (%d > %d)", b.config.ExpiryBuffer, flow.DefaultTransactionExpiry)
+	}
+
+	return &b, nil
 }
 
 // BuildOn creates a new block built on the given parent. It produces a payload
 // that is valid with respect to the un-finalized chain it extends.
 func (b *Builder) BuildOn(parentID flow.Identifier, setter func(*flow.Header) error) (*flow.Header, error) {
-	var proposal cluster.Block
+	var proposal cluster.Block                 // proposal we are building
+	var parent flow.Header                     // parent of the proposal we are building
+	var clusterChainFinalizedBlock flow.Header // finalized block on the cluster chain
+	var refChainFinalizedHeight uint64         // finalized height on reference chain
+	var refChainFinalizedID flow.Identifier    // finalized block ID on reference chain
 
 	startTime := time.Now()
 
-	// first we construct a proposal in-memory, ensuring it is a valid extension
-	// of chain state -- this can be done in a read-only transaction
-	err := b.db.View(func(tx *badger.Txn) error {
+	// STEP ONE: build a lookup for excluding duplicated transactions.
+	// This is briefly how it works:
+	//
+	// Let E be the global transaction expiry.
+	// When incorporating a new collection C, with reference height R, we enforce
+	// that it contains only transactions with reference heights in [R,R+E).
+	// * if we are building C:
+	//   * we don't build expired collections (ie. our local finalized consensus height is at most R+E-1)
+	//   * we don't include transactions referencing un-finalized blocks
+	//   * therefore, C will contain only transactions with reference heights in [R,R+E)
+	// * if we are validating C:
+	//   * honest validators only consider C valid if all its transactions have reference heights in [R,R+E)
+	//
+	// Therefore, to check for duplicates, we only need a lookup for transactions in collection
+	// with expiry windows that overlap with our collection under construction.
+	//
+	// A collection with overlapping expiry window can be finalized or un-finalized.
+	// * to find all non-expired and finalized collections, we make use of an index
+	//   (main_chain_finalized_height -> cluster_block_id), to search for a range of main chain heights
```

Comment:
the index returns a _set_ of blocks for a given `main_chain_finalized_height`

```suggestion
	//   (main_chain_finalized_height -> cluster_block_ids with respective reference height), to search for a range of main chain heights
```